Using local and s3 storage in docker compose setup

Hi!

When using local storage, I encounter the error:

```bash
Cannot write file
```

At the same time, there are no logs in docker-compose, which makes it difficult to understand the root cause of the issue.

Here is my .env configuration for local storage:
```bash
# File storage  
# NOTE: In a Docker environment, the "local" strategy will not work out of the box since  
# multiple containers would need to have access to the same local disk.  
# You can use the "s3" strategy instead or set up a shared volume and point the public files path to it.  
DRIVE_DISK=local # options: local, s3  

# Paths for local storage (optional)  
FILE_STORAGE_ROOT_PATH="/app/storage/files"  
FILE_PUBLIC_PATH="files"  
FILES_STORAGE_PATH="/app/storage/files"  
PUBLIC_FILES_STORAGE_PATH="/app/apps/web/public/files" # Must be inside Next.js public folder 
```

Here is the relevant part of my docker-compose.yml:

```yaml
x-common-service: &common-service
  env_file:
    - .env
  volumes:
    - shared-storage:/app/storage/files
    - shared-storage:/app/apps/web/public/files
  depends_on:
    - db
    - redis
  networks:
    - web

x-web-common: &web-common
  <<: *common-service
  labels:
    - "traefik.enable=true"
    - "traefik.docker.network=web"
    - "traefik.http.services.web-service.loadbalancer.server.port=8080"
    - "traefik.http.routers.web-http.entrypoints=http"
    - "traefik.http.routers.web-http.service=web-service"
    - "traefik.http.routers.web-http.rule=Host(`app-latitude.${APP_DOMAIN}`)"
    - "traefik.http.routers.web-https.tls.certresolver=letsencryptresolver"
    - "traefik.http.middlewares.web-redirect.redirectscheme.scheme=https"
    - "traefik.http.routers.web-http.middlewares=web-redirect"
    - "traefik.http.routers.web-https.entrypoints=https"
    - "traefik.http.routers.web-https.tls=true"
    - "traefik.http.routers.web-https.rule=Host(`app-latitude.${APP_DOMAIN}`)"
    - "traefik.http.routers.web-https.service=web-service"
  depends_on:
    - db
    - redis

x-gateway-common: &gateway-common
  <<: *common-service
  labels:
    - "traefik.enable=true"
    - "traefik.docker.network=web"
    - "traefik.http.services.gateway-service.loadbalancer.server.port=8080"
    - "traefik.http.routers.gateway-http.entrypoints=http"
    - "traefik.http.routers.gateway-http.service=gateway-service"
    - "traefik.http.routers.gateway-http.rule=Host(`gateway-latitude.${APP_DOMAIN}`)"
    - "traefik.http.routers.gateway-https.tls.certresolver=letsencryptresolver"
    - "traefik.http.middlewares.gateway-redirect.redirectscheme.scheme=https"
    - "traefik.http.routers.gateway-http.middlewares=gateway-redirect"
    - "traefik.http.routers.gateway-https.entrypoints=https"
    - "traefik.http.routers.gateway-https.tls=true"
    - "traefik.http.routers.gateway-https.rule=Host(`gateway-latitude.${APP_DOMAIN}`)"
    - "traefik.http.routers.gateway-https.service=gateway-service"

x-websockets-common: &websockets-common
  <<: *common-service
  labels:
    - "traefik.enable=true"
    - "traefik.docker.network=web"
    - "traefik.http.services.websockets-service.loadbalancer.server.port=4002"
    - "traefik.http.routers.websockets-http.entrypoints=http"
    - "traefik.http.routers.websockets-http.service=websockets-service"
    - "traefik.http.routers.websockets-http.rule=Host(`ws-latitude.${APP_DOMAIN}`)"
    - "traefik.http.routers.websockets-https.tls.certresolver=letsencryptresolver"
    - "traefik.http.middlewares.websockets-redirect.redirectscheme.scheme=https"
    - "traefik.http.routers.websockets-http.middlewares=websockets-redirect"
    - "traefik.http.routers.websockets-https.entrypoints=https"
    - "traefik.http.routers.websockets-https.tls=true"
    - "traefik.http.routers.websockets-https.rule=Host(`ws-latitude.${APP_DOMAIN}`)"
    - "traefik.http.routers.websockets-https.service=websockets-service"

services:
  mailpit:
    image: axllent/mailpit
    restart: always
    ports:
      - 8025:8025
      - 1025:1025
    profiles:
      - development

  db:
    image: postgres:15.8
    restart: always
    container_name: latitude-llm-db
    ports:
      - '5432:5432'
    env_file:
      - .env
    volumes:
      - ./docker/init-db.sh:/docker-entrypoint-initdb.d/init-db.sh
      - pgdata:/var/lib/postgresql/data
    networks:
    - web

  redis:
    image: redis
    restart: always
    container_name: latitude-llm-redis
    ports:
      - '6379:6379'
    networks:
    - web

  web:
    <<: *web-common
    image: ghcr.io/latitude-dev/web:latest
    restart: always
    container_name: latitude-llm-web
    platform: linux/amd64

  migrations:
    <<: *common-service
    image: ghcr.io/latitude-dev/migrations:latest
    container_name: latitude-llm-migrations
    platform: linux/amd64
    depends_on:
      - db

  gateway:
    <<: *gateway-common
    image: ghcr.io/latitude-dev/gateway:latest
    restart: always
    container_name: latitude-llm-gateway
    platform: linux/amd64

  workers:
    <<: *common-service
    image: ghcr.io/latitude-dev/workers:latest
    restart: always
    container_name: latitude-llm-workers
    platform: linux/amd64

  websockets:
    <<: *websockets-common
    image: ghcr.io/latitude-dev/websockets:latest
    restart: always
    container_name: latitude-llm-websockets
    platform: linux/amd64

networks:
  web:
    external: true

volumes:
  pgdata:
  shared-storage:
```

Despite mounting shared-storage as a volume, the issue persists.

Could you help me understand why this error occurs and how to resolve it?

### Question about S3 authentication

I understand that I can use AWS_ACCESS_KEY and AWS_ACCESS_SECRET to connect to an S3 bucket. However, in some cases, I don't want to issue separate credentials for writing to the bucket. Instead, I would like to use IAM roles that I can assign to an EC2 instance.

Would it be possible to add support for using IAM roles instead of access keys for authentication? That would be a great improvement!

Thanks!

I need clarification on which containers use the following paths and how relevant they are:

```bash
FILE_STORAGE_ROOT_PATH="/app/storage/files"  
FILE_PUBLIC_PATH="files"  
FILES_STORAGE_PATH="/app/storage/files"  
PUBLIC_FILES_STORAGE_PATH="/app/apps/web/public/files" # Must be inside the Next.js public folder
```

This information is necessary for me to make and test changes in Dockerfile, docker-compose.yml, and the documentation.

Additionally, when starting the containers, I encountered the following error:

```bash
latitude-llm-gateway     | Warning: Cannot load "@napi-rs/canvas" package: "Error: Cannot find module '@napi-rs/canvas'
latitude-llm-gateway     | Require stack:
latitude-llm-gateway     | - /app/apps/gateway/dist/server.js".
latitude-llm-gateway     | Warning: Cannot polyfill `DOMMatrix`, rendering may be broken.
latitude-llm-gateway     | Warning: Cannot polyfill `ImageData`, rendering may be broken.
latitude-llm-gateway     | Warning: Cannot polyfill `Path2D`, rendering may be broken.
latitude-llm-gateway     | Listening on http://0.0.0.0:8080
```

Could you please advise on how to resolve this?

--- a/.env.example
+++ b/.env.example
@@ -55,11 +55,13 @@ SENTRY_PROJECT=
 DRIVE_DISK=local # options: local, s3
 
 # Paths for local storage (optional)
-FILE_PUBLIC_PATH= # e.g files
-FILES_STORAGE_PATH= # e.g /app/storage/files
-PUBLIC_FILES_STORAGE_PATH= # e.g /app/storage/public/files (IMPORTANT: has to be in nextjs's public folder)
+FILE_STORAGE_ROOT_PATH="/app/storage/files" # e.g /app/storage/files
+FILE_PUBLIC_PATH="files" # e.g files
+FILES_STORAGE_PATH="/app/storage/files" # e.g /app/storage/files
+PUBLIC_FILES_STORAGE_PATH="/app/apps/web/public/files" # e.g /app/storage/public/files (IMPORTANT: has to be in nextjs's public folder)
 
 # AWS S3 (optional)
+# If you are using AWS IAM roles, you can skip AWS_ACCESS_KEY and AWS_ACCESS_SECRET
 ASW_REGION=
 AWS_ACCESS_KEY=
 AWS_ACCESS_SECRET=

--- a/apps/gateway/docker/Dockerfile
+++ b/apps/gateway/docker/Dockerfile
@@ -87,6 +87,7 @@ RUN apk add --no-cache \
 RUN addgroup --system --gid 1001 nodejs
 RUN adduser --system --uid 1001 latitude
 # Set permissions for local storage
+# User ID and Group ID 1001 is used to match the user 'latitude' and group 'nodejs' in the builder image.
 RUN set -e; \
     mkdir -p /app/storage/files; \
     mkdir -p /app/apps/web/public/files; \

--- a/apps/gateway/docker/Dockerfile
+++ b/apps/gateway/docker/Dockerfile
@@ -86,6 +86,11 @@ RUN apk add --no-cache \
 
 RUN addgroup --system --gid 1001 nodejs
 RUN adduser --system --uid 1001 latitude
+# Set permissions for local storage
+RUN set -e; \
+    mkdir -p /app/storage/files; \
+    mkdir -p /app/apps/web/public/files; \
+    chown -R 1001:1001 /app/storage/files /app/apps/web/public/files
 USER latitude
 
 WORKDIR /app

--- a/apps/web/docker/Dockerfile
+++ b/apps/web/docker/Dockerfile
@@ -157,6 +157,7 @@ COPY --from=builder --chown=nodejs:nextjs /app/apps/web/.next/static ./apps/web/
 COPY --from=builder --chown=nodejs:nextjs /app/node_modules ./node_modules
 COPY --from=builder --chown=nodejs:nextjs /app/apps/web/node_modules ./apps/web/node_modules
 # Set permissions for local storage
+# User ID and Group ID 1001 is used to match the user 'nextjs' and group 'nodejs' in the runner image.
 RUN set -e; \
     mkdir -p /app/storage/files; \
     mkdir -p /app/apps/web/public/files; \

--- a/apps/web/docker/Dockerfile
+++ b/apps/web/docker/Dockerfile
@@ -156,6 +156,11 @@ COPY --from=builder --chown=nodejs:nextjs /app/apps/web/.next/static ./apps/web/
 # Copy all node_modules to ensure dependencies are available
 COPY --from=builder --chown=nodejs:nextjs /app/node_modules ./node_modules
 COPY --from=builder --chown=nodejs:nextjs /app/apps/web/node_modules ./apps/web/node_modules
+# Set permissions for local storage
+RUN set -e; \
+    mkdir -p /app/storage/files; \
+    mkdir -p /app/apps/web/public/files; \
+    chown -R 1001:1001 /app/storage/files /app/apps/web/public/files
 
 USER nextjs
 

--- a/apps/workers/docker/Dockerfile
+++ b/apps/workers/docker/Dockerfile
@@ -74,6 +74,11 @@ RUN apk add --no-cache \
 
 RUN addgroup --system --gid 1001 nodejs
 RUN adduser --system --uid 1001 latitude
+# Set permissions for local storage
+RUN set -e; \
+    mkdir -p /app/storage/files; \
+    mkdir -p /app/apps/web/public/files; \
+    chown -R 1001:1001 /app/storage/files /app/apps/web/public/files
 USER latitude
 
 WORKDIR /app

--- a/apps/workers/docker/Dockerfile
+++ b/apps/workers/docker/Dockerfile
@@ -75,6 +75,7 @@ RUN apk add --no-cache \
 RUN addgroup --system --gid 1001 nodejs
 RUN adduser --system --uid 1001 latitude
 # Set permissions for local storage
+# User ID and Group ID 1001 is used to match the user 'latitude' and group 'nodejs' in the runner image.
 RUN set -e; \
     mkdir -p /app/storage/files; \
     mkdir -p /app/apps/web/public/files; \

--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -171,8 +171,6 @@ services:
     restart: always
     container_name: latitude-llm-web
     platform: linux/amd64
-    profiles:
-      - default
 
   migrations-local:
     <<: *common-service
@@ -196,8 +194,6 @@ services:
       - "traefik.enable=false"
     depends_on:
       - db
-    profiles:
-      - default
 
   gateway-local:
     <<: *gateway-common
@@ -214,8 +210,6 @@ services:
     restart: always
     container_name: latitude-llm-gateway
     platform: linux/amd64
-    profiles:
-      - default
 
   workers-local:
     <<: *common-service
@@ -232,8 +226,6 @@ services:
     restart: always
     container_name: latitude-llm-workers
     platform: linux/amd64
-    profiles:
-      - default
 
   websockets-local:
     <<: *websockets-common
@@ -250,8 +242,6 @@ services:
     restart: always
     container_name: latitude-llm-websockets
     platform: linux/amd64
-    profiles:
-      - default
 
 # You need to create external network for Traefik to work (docker network create web)
 networks:

--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -1,6 +1,9 @@
 x-common-service: &common-service
   env_file:
     - .env
+  volumes:
+    - shared-storage:/app/storage/files:rw
+    - shared-storage:/app/apps/web/public/files:rw
   depends_on:
     - db
     - redis
@@ -168,6 +171,8 @@ services:
     restart: always
     container_name: latitude-llm-web
     platform: linux/amd64
+    profiles:
+      - default
 
   migrations-local:
     <<: *common-service
@@ -175,7 +180,8 @@ services:
     build:
       context: .
       dockerfile: packages/core/docker/Dockerfile
-
+    labels:
+      - "traefik.enable=false"
     depends_on:
       - db
     profiles:
@@ -190,6 +196,8 @@ services:
       - "traefik.enable=false"
     depends_on:
       - db
+    profiles:
+      - default
 
   gateway-local:
     <<: *gateway-common
@@ -197,7 +205,6 @@ services:
     build:
       context: .
       dockerfile: apps/gateway/docker/Dockerfile
-
     profiles:
       - local
 
@@ -207,6 +214,8 @@ services:
     restart: always
     container_name: latitude-llm-gateway
     platform: linux/amd64
+    profiles:
+      - default
 
   workers-local:
     <<: *common-service
@@ -214,7 +223,6 @@ services:
     build:
       context: .
       dockerfile: apps/workers/docker/Dockerfile
-
     profiles:
       - local
 
@@ -224,6 +232,8 @@ services:
     restart: always
     container_name: latitude-llm-workers
     platform: linux/amd64
+    profiles:
+      - default
 
   websockets-local:
     <<: *websockets-common
@@ -231,7 +241,6 @@ services:
     build:
       context: .
       dockerfile: apps/websockets/docker/Dockerfile
-
     profiles:
       - local
 
@@ -241,8 +250,13 @@ services:
     restart: always
     container_name: latitude-llm-websockets
     platform: linux/amd64
+    profiles:
+      - default
 
 # You need to create external network for Traefik to work (docker network create web)
 networks:
   web:
     external: true
+
+volumes:
+  shared-storage:

--- a/docs/guides/self-hosted/production.mdx
+++ b/docs/guides/self-hosted/production.mdx
@@ -71,7 +71,67 @@ Make sure this network is created before running the containers with  `docker co
 - **Storage Configuration**:
 
   - `DRIVE_DISK`: Choose between `local` or `s3` for file storage
-  - AWS S3 credentials (if using S3 storage)
+    1. `local` file storage configuration:
+      - Files are stored locally on the host machine using Docker volumes.
+      - Default variables used:
+
+        ```bash
+        # Paths for local storage (optional)
+        FILE_STORAGE_ROOT_PATH="/app/storage/files" # e.g /app/storage/files
+        FILE_PUBLIC_PATH="files" # e.g files
+        FILES_STORAGE_PATH="/app/storage/files" # e.g /app/storage/files
+        PUBLIC_FILES_STORAGE_PATH="/app/apps/web/public/files" # e.g /app/storage/public/files (IMPORTANT: has to be in nextjs's public folder)
+        ```
+
+    2. `s3` AWS S3 storage configuration:
+      - With environment variables (for convenience/legacy):
+        You explicitly provide AWS credentials (AWS_ACCESS_KEY and AWS_ACCESS_SECRET) via .env file.
+        Required variables:
+
+        ```bash
+        AWS_ACCESS_KEY=your-access-key-here
+        AWS_ACCESS_SECRET=your-secret-here
+        AWS_REGION=us-east-1
+        S3_BUCKET=app-files-bucket
+        PUBLIC_S3_BUCKET=public-files-bucket
+        ```
+      - AWS S3 with IAM Roles (recommended):
+        No explicit AWS keys needed!
+        Use IAM Roles attached to your AWS services (ECS, EC2, Lambda).
+        Ensure AWS resource has proper IAM Role with S3 access (GetObject, PutObject, DeleteObject).
+        Only required environment variables (no keys explicitly stored):
+
+        ```bash
+        AWS_REGION=us-east-1
+        S3_BUCKET=app-files-bucket
+        PUBLIC_S3_BUCKET=public-files-bucket
+        ```
+        How to configure IAM Role (example):
+        1. Go to AWS IAM → Create Role.
+        2. Select trusted entity type (e.g., AWS Service).
+        3. Attach policy that allows access to required S3 buckets.
+
+        ```json
+        {
+          "Version": "2012-10-17",
+          "Statement": [{
+            "Effect": "Allow",
+            "Action": [
+              "s3:GetObject",
+              "s3:PutObject",
+              "s3:DeleteObject",
+              "s3:ListBucket"
+            ],
+            "Resource": [
+              "arn:aws:s3:::your-bucket-name-here",
+              "arn:aws:s3:::your-bucket-name-here/*"
+            ]
+          }]
+        }
+        ```
+        4. Attach this IAM role to AWS infrastructure (EC2 Instance / ECS Task Definition / Lambda function).
+
+        AWS SDK automatically handles credentials from attached IAM roles.
 
 - **Optional Features**:
   - Sentry integration for error tracking

--- a/packages/core/src/lib/disk.ts
+++ b/packages/core/src/lib/disk.ts
@@ -27,12 +27,11 @@ function getAwsConfig() {
   const bucket = env.S3_BUCKET
   const publicBucket = env.PUBLIC_S3_BUCKET
   const region = env.AWS_REGION
-  
-  if (!bucket || !publicBucket || !region) {
+
+  if (!bucket || !publicBucket || !region)
     throw new Error(
-      '(PUBLIC)_S3_BUCKET and AWS_REGION variables are required.',
+      `Missing required AWS configuration variables: ${[!bucket && 'S3_BUCKET', !publicBucket && 'PUBLIC_S3_BUCKET', !region && 'AWS_REGION'].filter(Boolean).join(', ')}.`,
     )
-  }
 
   const accessKeyId = env.AWS_ACCESS_KEY
   const secretAccessKey = env.AWS_ACCESS_SECRET

--- a/packages/core/src/lib/disk.ts
+++ b/packages/core/src/lib/disk.ts
@@ -109,26 +109,20 @@ export class DiskWrapper {
 
   async putStream(key: string, contents: Readable, options?: WriteOptions) {
     try {
-      console.log(`Uploading file: key=${key}, contentLength=${contents.readableLength}`)
-  
       await this.disk.putStream(key, contents, {
         ...options,
         contentLength: contents.readableLength,
       })
-  
-      console.log(`Successfully uploaded: key=${key}`)
       return Result.nil()
     } catch (e) {
-      console.error(`Error uploading file: key=${key}`, e)
-  
       if (e instanceof errors.E_CANNOT_WRITE_FILE) {
         return Result.error(new Error('Cannot write file'))
       }
-  
-      return Result.error(e as Error)
+
+      const error = e as Error
+      return Result.error(error)
     }
   }
-  
 
   async delete(key: string | null | undefined) {
     if (!key) return Result.nil()

--- a/packages/core/src/lib/disk.ts
+++ b/packages/core/src/lib/disk.ts
@@ -38,7 +38,6 @@ function getAwsConfig() {
   const secretAccessKey = env.AWS_ACCESS_SECRET
 
   if (accessKeyId && secretAccessKey) {
-    // Используем заданные ключи доступа
     return {
       region,
       bucket,
@@ -47,12 +46,11 @@ function getAwsConfig() {
     }
   }
 
-  // Если ключи не заданы — подразумеваем IAM-роль
+  // If AWS_ACCESS_KEY and AWS_ACCESS_SECRET are not set, the SDK will use AWS IAM role.
   return {
     region,
     bucket,
     publicBucket,
-    // не передаем credentials
   }
 }
 

--- a/packages/core/src/lib/disk.ts
+++ b/packages/core/src/lib/disk.ts
@@ -24,23 +24,35 @@ const generateUrl =
  * AWS_ACCESS_SECRET=[your-secret]
  */
 function getAwsConfig() {
-  const accessKeyId = env.AWS_ACCESS_KEY
   const bucket = env.S3_BUCKET
   const publicBucket = env.PUBLIC_S3_BUCKET
   const region = env.AWS_REGION
-  const secretAccessKey = env.AWS_ACCESS_SECRET
-
-  if (!accessKeyId || !secretAccessKey || !bucket || !publicBucket || !region) {
+  
+  if (!bucket || !publicBucket || !region) {
     throw new Error(
-      'AWS credentials not configured. Check you setup AWS_ACCESS_KEY, AWS_ACCESS_SECRET, (PUBLIC)_S3_BUCKET and AWS_REGION in your .env file.',
+      '(PUBLIC)_S3_BUCKET and AWS_REGION variables are required.',
     )
   }
 
+  const accessKeyId = env.AWS_ACCESS_KEY
+  const secretAccessKey = env.AWS_ACCESS_SECRET
+
+  if (accessKeyId && secretAccessKey) {
+    // Используем заданные ключи доступа
+    return {
+      region,
+      bucket,
+      publicBucket,
+      credentials: { accessKeyId, secretAccessKey },
+    }
+  }
+
+  // Если ключи не заданы — подразумеваем IAM-роль
   return {
     region,
     bucket,
     publicBucket,
-    credentials: { accessKeyId, secretAccessKey },
+    // не передаем credentials
   }
 }
 
@@ -97,20 +109,26 @@ export class DiskWrapper {
 
   async putStream(key: string, contents: Readable, options?: WriteOptions) {
     try {
+      console.log(`Uploading file: key=${key}, contentLength=${contents.readableLength}`)
+  
       await this.disk.putStream(key, contents, {
         ...options,
         contentLength: contents.readableLength,
       })
+  
+      console.log(`Successfully uploaded: key=${key}`)
       return Result.nil()
     } catch (e) {
+      console.error(`Error uploading file: key=${key}`, e)
+  
       if (e instanceof errors.E_CANNOT_WRITE_FILE) {
         return Result.error(new Error('Cannot write file'))
       }
-
-      const error = e as Error
-      return Result.error(error)
+  
+      return Result.error(e as Error)
     }
   }
+  
 
   async delete(key: string | null | undefined) {
     if (!key) return Result.nil()

